### DISCUSSION TITLE
Layer example

### DISCUSSION THREAD

**Discussion by v0id-4lpz** [nuxt/nuxt#26634] [General]
Hey guys!

I'm stuck and I really need help.

I try to create reusable codebase with some features (auth, admin, components...) in separates privates GitHub repos.

- NuxtApp1 extends NuxtBaseLayer
- NuxtApp2 extends NuxtBaseLayer, NuxtExtraLayer

I started to create a layer but idk why but i'm not able to do it work...

I read all the tutos on google, watched all videos on YouTube but nothing work...
All the time a problem
- can't load layer
- can't load module
- dependencie not found
- ...

No one say the same way on config... I'm really lost and stressed about that.

As I can see I'm not the only one that have issues with layers.
Please, can someone explain us more or give us a step-by-step working tutorial?

Thank you :*

**Comment by v0id-4lpz** [comment]
Here is the state of my simples tests

node 20.11.0
npm 10.2.4


## Setup app

```
mkdir nuxt-env
cd /nuxt-env

npx nuxi@latest init myApp
```


## Local layer

```
npx nuxi init --template layer myLayer

cd /nuxt-env/myLayer
npm install moment
```

`/nuxt-env/myApp/app.vue`
```
<template>
  <div>
    {{ date }}
  </div>
</template>

<script setup lang="ts">
import moment from "moment"

const date = moment()
</script>
```

`/nuxt-env/myApp/nuxt.config.ts`
```
extends: ['../myLayer']
```

```
cd /nuxt-env/myApp
npm run dev

> 500 moment is not defined
```


## Repo layer (Github)

This is the same layer as previous "local layer" but pushed on a Github repo

`/nuxt-env/myApp/nuxt.config.ts`
```
extends: [
	['github:myAccount/myLayerRepo#master', { install: true, auth: 'my_github_token' }]
]
```

```
cd /nuxt-env/myApp
npm install

...downloading layer...

>  ERROR  Cannot find module 'D:/Projects/nuxt-env/app/moment'
```

```
cd /nuxt-env/myApp
npm run dev

>  ERROR  Cannot find module 'D:/Projects/nuxt-env/app/moment'
```


## npm package layer

Don't want to deal with npm packages

**Comment by v0id-4lpz** [comment]
up

I took time to dev my layer and it's working... BUT only if I manually install it's deps on each of my projects...
Still looking for a clean way to not install em manually.

**Comment by madsh93** [comment]
Can you post your respective package.json and Nuxt configs?

**Reply by v0id-4lpz** [reply]
There is default packages.json except for layer where I installed `moment` for my test :)

EDIT: sorry i missed to past my nuxt.config

## App

`/nuxt-env/myApp/package.json` (default)
```json
{
  "name": "app",
  "private": true,
  "type": "module",
  "scripts": {
    "build": "nuxt build",
    "dev": "nuxt dev",
    "generate": "nuxt generate",
    "preview": "nuxt preview",
    "postinstall": "nuxt prepare"
  },
  "dependencies": {
    "nuxt": "^3.11.2",
    "vue": "^3.4.21",
    "vue-router": "^4.3.0"
  }
}
```

`/nuxt-env/myApp/nuxt.config.ts`
```ts
export default defineNuxtConfig({
	devtools: { enabled: true },
	extends: [
		//"../layer",
		["github:myAccount/myLayer#master", { install: true, auth: "github_token" }],
	],
});
```

## Layer

`/nuxt-env/myLayer/package.json`
```json
{
  "name": "my-nuxt-layer",
  "type": "module",
  "version": "0.0.1",
  "main": "./nuxt.config.ts",
  "scripts": {
    "dev": "nuxi dev .playground",
    "build": "nuxt build .playground",
    "generate": "nuxt generate .playground",
    "preview": "nuxt preview .playground",
    "lint": "eslint .",
    "postinstall": "nuxt prepare .playground"
  },
  "devDependencies": {
    "@nuxt/eslint-config": "^0.1.1",
    "eslint": "^8.28.0",
    "nuxt": "^3.6.2",
    "typescript": "^4.9.3"
  },
  "dependencies": {
    "moment": "^2.30.1"
  }
}
```

`/nuxt-env/myLayer/nuxt.config.ts` (default)
```ts
// https://nuxt.com/docs/api/configuration/nuxt-config
export default defineNuxtConfig({
  devtools: { enabled: true }
})

```

**Comment by v0id-4lpz** [comment]
When I link my layer from Github repo and `npm install` it will well downloaded and installed in `/myApp/node_modules/.c12/github_xxx_xxx`.

Node modules of my layer are installed in `/myApp/node_modules/.c12/github_repo_xxx/node_modules` but unusables.

Something seems missing in my config or in Nuxt/c12

**Comment by Sreesanth46** [comment]
+1
